binder/src/main/java/io/grpc/binder/internal/BinderClientTransport.java

```diff
@@ -316,14 +316,21 @@ void notifyTerminated() {
       readyTimeoutFuture.cancel(false);
       readyTimeoutFuture = null;
     }
-    if (preAuthResultFuture != null) {
-      preAuthResultFuture.cancel(false); // No effect if already complete.
-    }
-    if (authResultFuture != null) {
-      authResultFuture.cancel(false); // No effect if already complete.
-    }
+
+    ListenableFuture<Status> preAuthFuture = preAuthResultFuture;
+    ListenableFuture<Status> authFuture = authResultFuture;
+    preAuthResultFuture = null;
+    authResultFuture = null;
+
     serviceBinding.unbind();
     clientTransportListener.transportTerminated();
+
+    if (preAuthFuture != null) {
```

Comment:
99.9% of the time these futures will already be done at this point. Now that cancellation is more expensive, can make the cancel() conditional on `!isDone()`. There's no new race here because a done future will always stay done. If not done, it already races with cancel() but that's the nature of cancellation.